rasa/nlu/extractors/entity_synonyms.py

```diff
@@ -1,102 +1,142 @@
+
+        if self._resource:
+            self._persist()
+
+    def process(self, messages: List[Message]) -> List[Message]:
+        """Modifies entities attached to message to resolve synonyms.
 
-    def process(self, message: Message, **kwargs: Any) -> None:
+        Args:
+            messages: List containing the latest user message
 
-        updated_entities = message.get(ENTITIES, [])[:]
-        self.replace_synonyms(updated_entities)
-        message.set(ENTITIES, updated_entities, add_to_output=True)
+        Returns:
+            List containing the latest user message with entities resolved to
+            synonyms if there is a match.
+        """
+        for message in messages:
+            updated_entities = message.get(ENTITIES, [])[:]
+            self.replace_synonyms(updated_entities)
+            message.set(ENTITIES, updated_entities, add_to_output=True)
 
-    def persist(self, file_name: Text, model_dir: Text) -> Optional[Dict[Text, Any]]:
+        return messages
+
+    def _persist(self) -> None:
 
         if self.synonyms:
-            file_name = file_name + ".json"
-            entity_synonyms_file = os.path.join(model_dir, file_name)
-            write_json_to_file(
-                entity_synonyms_file, self.synonyms, separators=(",", ": ")
-            )
-            return {"file": file_name}
-        else:
-            return {"file": None}
+            with self._model_storage.write_to(self._resource) as storage:
+                entity_synonyms_file = (
+                    storage / EntitySynonymMapperComponent.SYNONYM_FILENAME
+                )
 
+                write_json_to_file(
+                    entity_synonyms_file, self.synonyms, separators=(",", ": ")
+                )
+
+    # Adapt to get path from model storage and resource
     @classmethod
     def load(
         cls,
-        meta: Dict[Text, Any],
-        model_dir: Text,
-        model_metadata: Optional[Metadata] = None,
-        cached_component: Optional["EntitySynonymMapper"] = None,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
         **kwargs: Any,
-    ) -> "EntitySynonymMapper":
+    ) -> EntitySynonymMapperComponent:
         """Loads trained component (see parent class for full docstring)."""
-        file_name = meta.get("file")
-        if not file_name:
-            synonyms = None
-            return cls(meta, synonyms)
-
-        entity_synonyms_file = os.path.join(model_dir, file_name)
-        if os.path.isfile(entity_synonyms_file):
-            synonyms = rasa.shared.utils.io.read_json_file(entity_synonyms_file)
-        else:
-            synonyms = None
-            rasa.shared.utils.io.raise_warning(
-                f"Failed to load synonyms file from '{entity_synonyms_file}'.",
-                docs=DOCS_URL_TRAINING_DATA + "#synonyms",
+        with model_storage.read_from(resource) as storage:
```

Comment:
this might throw a `ValueError` in case their is no persisted data for this `Resource`. Can we please handle that (e.g. see how the other components do it)?